Release 4.0.0 (#1506)
microsoft/onefuzz: CHANGELOG.md

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 4.0.0
### Added
* Agent: Added common source coverage format. [#1403](https://github.com/microsoft/onefuzz/pull/1403)
* Service: Added class to store and retrieve rules associated with an API endpoint. This supports the ability to control who has access to an API. [#1420](https://github.com/microsoft/onefuzz/pull/1420)

> * Service: Support for creation of NSG during deployment allowing restricted access to the scaleset and repro VM's. [#1331](https://github.com/microsoft/onefuzz/pull/1331), [#1340](https://github.com/microsoft/onefuzz/pull/1340), [#1358](https://github.com/microsoft/onefuzz/pull/1358), [#1385](https://github.com/microsoft/onefuzz/pull/1385), [#1393](https://github.com/microsoft/onefuzz/pull/1393), [#1395](https://github.com/microsoft/onefuzz/pull/1395), [#1400](https://github.com/microsoft/onefuzz/pull/1400), [#1404](https://github.com/microsoft/onefuzz/pull/1404), [#1406](https://github.com/microsoft/onefuzz/pull/1406), [#1410](https://github.com/microsoft/onefuzz/pull/1410)

Reviewer: ```suggestion
* Service: Support for NSG creation during deployment, allowing restricted access to the scaleset and repro VMs. [#1331](https://github.com/microsoft/onefuzz/pull/1331), [#1340](https://github.com/microsoft/onefuzz/pull/1340), [#1358](https://github.com/microsoft/onefuzz/pull/1358), [#1385](https://github.com/microsoft/onefuzz/pull/1385), [#1393](https://github.com/microsoft/onefuzz/pull/1393), [#1395](https://github.com/microsoft/onefuzz/pull/1395), [#1400](https://github.com/microsoft/onefuzz/pull/1400), [#1404](https://github.com/microsoft/onefuzz/pull/1404), [#1406](https://github.com/microsoft/onefuzz/pull/1406), [#1410](https://github.com/microsoft/onefuzz/pull/1410)
```

> * Service: Disable guest account access by default when creating the default service principal during deployment. [#1425](https://github.com/microsoft/onefuzz/pull/1425)

Reviewer: ```suggestion
* Service: Disabled guest account access by default when creating the default service principal during deployment. [#1425](https://github.com/microsoft/onefuzz/pull/1425)
```

Author: The past tense doesn't sound right to me.

Reviewer: I suggested this since we use past tense _everywhere_ else in this changelog, but I think you're right in this case, given how the verb is being used (we're describing a policy, vs. a past-tense change to the code like "added" or "fixed").

Author: Maybe "Guest account access disabled by default..."?

> * Service: Added support for compute gallery images. [#1450](https://github.com/microsoft/onefuzz/pull/1450)

Reviewer: I think this should be in "Fixed", since we already claimed to support it in our docs.

Author: The gallery images?

Reviewer: Yep. From the PR:
> The documentation for using custom images while fuzzing states that Compute Gallery images are supported.

We should also capitalize "Compute Gallery".

* Service: Group membership check added. [#1074](https://github.com/microsoft/onefuzz/pull/1074)

> * Service: Exposed the `target_timeout` parameter in the radamsa template. [#1499](https://github.com/microsoft/onefuzz/pull/1499)

Reviewer: ```suggestion
* Service: Exposed the `target_timeout` parameter in the `radamsa basic` template. [#1499](https://github.com/microsoft/onefuzz/pull/1499)
```

### Fixed

> * Service: Azure DevOps was missing connect call which prevented connecting in some cases. [#1370](https://github.com/microsoft/onefuzz/pull/1370)

Reviewer: ```suggestion
* Service: Fix Azure DevOps work item creation by adding missing client initialization. [#1370](https://github.com/microsoft/onefuzz/pull/1370)
```

> * Service: Fix Azure DevOps work item creation by adding missing client initialization. [#1370](https://github.com/microsoft/onefuzz/pull/1370)

Reviewer 2: Change to "Fixed"

> * Service: Fixed validation of the `target_exe` blob name preventing spurious failures when nested in a (virtual) directory. [#1371](https://github.com/microsoft/onefuzz/pull/1371)

Reviewer: ```suggestion
* Service: Fixed validation of the `target_exe` blob name, enabling nesting in a subdirectory of the `setup` container. [#1371](https://github.com/microsoft/onefuzz/pull/1371)
```

> * Service: Migrated to msgraph as `azure-graphrbac` is soon to be depreciated. [#966](https://github.com/microsoft/onefuzz/pull/966)

Reviewer: ```suggestion
* Service: Migrated to MS Graph, as `azure-graphrbac` is soon to be deprecated. [#966](https://github.com/microsoft/onefuzz/pull/966)
```

* Service: Stopped ignoring unexpected errors when authenticating the client secret. [#1376](https://github.com/microsoft/onefuzz/pull/1376)

> * Service: Fixed regex to correctly capture the object_id when trying to remove an invalid application id. [#1408](https://github.com/microsoft/onefuzz/pull/1408)

Reviewer: ```suggestion
* Service: Fixed regex to correctly capture the object ID when trying to remove an invalid application ID. [#1408](https://github.com/microsoft/onefuzz/pull/1408)
```
